proposed/swift-interop.md

```diff
@@ -0,0 +1,170 @@
+
+Swift also provides a strong Objective-C interop story through the `@objc` attribute. Types with the `@objc` attribute are exposed through the Swift ABI as well as through Objective-C selectors. We will not call into Swift APIs using the Objective-C interop in this interop story. However, the upstack projection tooling may need to do additional work for types that may need to have Objective-C interop support exposed (in the case that a .NET user type derives from a Swift type that had the `@objc` attribute).
+
+##### Self register
+
+We have two options for supporting the Self register in the calling convention:
+
+1. Use an attribute like `[SwiftSelf]` on a parameter to specify it should go in the self register.
+2. Specify that combining the `Swift` calling convention with the `MemberFunction` calling convention means that the first argument is the `self` argument.
+3. Use a `SwiftSelf<T>` argument to represent which parameter should go into the self register.
+
+The first option seems like a natural fit, but there is one significant limitation: Attributes cannot be used in function pointers. Function pointers are a vital scenario for us to support as we want to support virtual method tables as they are used in scenarios like protocol witness tables. The second option is a natural fit as the `MemberFunction` calling convention combined with the various C-based calling conventions specifies that there is a `this` argument as the first argument. Defining `Swift` + `MemberFunction` to imply/require the `self` argument is a great conceptual extension of the existing model.
+
+In Swift, sometimes the `self` register is used for non-instance state. For example, in static functions, the type metadata is passed as the `self` argument. Since static functions are not member functions, we may want to not use the `MemberFunction` calling convention. Alternatively, we could provide a `SwiftSelf<T>` type to specify "this argument goes in the self register". Specifying the type twice in a signature would generate an `InvalidProgramException`. This would allow the `self` argument to be specified anywhere in the argument list.
+
+For reference, explicitly declaring a function with the Swift or SwiftAsync calling conventions in Clang requires the "context" argument, the value that goes in the "self" register, as the last parameter or the penultimate parameter followed by the error parameter.
+
+###### Error register
+
+We have many options for handling the error register in the Swift calling convention:
+
+1. Use an attribute like `[SwiftError]` on a by-ref or `out` parameter to indicate the error parameter.
+2. Use a special type like `SwiftError` on a by-ref or `out` parameter to indicate the error parameter.
```

Comment:
ref/out parameters aren't blittable, so maybe should this be a `SwiftError*` instead?

Reply:
Since we don't actually pass the pointer value in the call (and instead use it basically as a storage location for the result), we can make the `ref/out` behavior not cause a problem due to blittability. We also may be able to optimize it better by using a managed ref instead of a pointer (as the address-exposure of the managed ref could be optimized away, but we can't do that for an unmanaged pointer). If we don't want to special case it this much though, an unmanaged pointer is also a reasonable option.

Reply:
Looks like C# disallows `ref/out` in `UnmanagedCallersOnly` signatures. I'll update the proposal with some alternative designs.